docs/core/v3.0.rst

```diff
@@ -757,21 +769,104 @@ elements in the order (0, 0), (1, 0), (2, 0), ..., (`dy` - 3, `dx` -
-consistent with the data type and memory layout of *a*.
+Chunks are encoded into a binary representation for storage in a store_, using
+the chain of codecs_ specified by the ``codecs`` metadata field.
+
+Determination of encoded representations
+----------------------------------------
+
+To encode or decode a chunk, the encoded and decoded representations for each
+codec in the chain must first be determined as follows:
+
+1. The initial decoded representation, ``decoded_representation[0]`` is
+   multi-dimensional array with the same data type as the zarr array, and a
+   shape determined according to the value of ``chunk_memory_layout`` as
+   follows:
+
+   - If ``chunk_memory_layout`` is equal to ``"C"``, the shape is equal to the
+     chunk shape.
+   - If ``chunk_memory_layout`` is equal to ``"F"``, the shape is equal to the
+     chunk shape, with the dimension order reversed.
+   - If ``chunk_memory_layout`` is defined by an extension, the extension
+     defines the shape.
+
+2. For each codec ``i``, the encoded representation is equal to the decoded
+   representation ``decoded_representation[i+1]`` of the next codec, and is
+   computed from
+   ``codecs[i].compute_encoded_representation_type(decoded_representation[i])``.
+   If ``compute_encoded_representation_type`` fails because of an incompatible
+   decoded representation, an implementation should indicate an error.
+
+.. _default-array-byte-string-conversion:
+
+Conversion between multi-dimensional array and byte string representations
+--------------------------------------------------------------------------
+
+Some codecs operate directly on multi-dimensional arrays of elements,
+e.g. encoding a 3-d array as a multi-channel jpeg image.  Other codecs operate
+at the byte level, e.g. gzip compression.  If a codec that operates at the byte
+level receives as input an array that is not a 1-dimensional uint8 array, it may
+convert the input array to a byte string by concatenating the default binary
+representations of each element in lexicographical order (C order).  Similarly,
+if a codec that expects a multi-dimensional array as input instead receives a
+byte string, it may decode each element in lexicographical order according to
+the default binary representation of each element.
+
+Encoding procedure
+------------------
+
+Based on the computed ``decoded_representations`` list, a chunk is encoded using
+the following procedure:
+
+1. The chunk array ``A`` (with a shape equal to the chunk shape, and data type
+   equal to the zarr array data type) is logically transformed into the initial
+   *encoded chunk* ``EC[0]`` of the type specified by
+   ``decoded_representation[0]`` according to the ``chunk_memory_layout`` as
+   follows:
+   
+   - If ``chunk_memory_layout`` is equal to ``"C"``, ``EC[0]`` equals ``A`` (no
+     transformation).
+   - If ``chunk_memory_layout`` is equal to ``"F"``, the dimension order is reversed.
```

Comment:
What does `the dimension order is reversed` mean here exactly? Would this need to change the binary format or just interpret it differently?
In the end, does this drop F-ordered data in the binary representation? I think that would be incovenient, since this means current F-ordered data cannot be read, or at least not with the dimension-order that was intended for the dataset.

Reply:
No, this doesn't drop support for F order.   The intent is that we describe the behavior of codecs in terms of logical arrays without concern for their memory layout.   Otherwise, I think it would be much more complicated to have to additionally specify for each codec how it affects the memory layout of the array, in addition to its logical value.  Since I have defined that when converting between a logical array and a byte sequence, the elements are always encoded in C order, to achieve F order we must transpose the dimensions first.  Actually there was some discussion previously of eliminating "chunk_memory_layout" and instead specifying it via a "transpose" codec; if that were done, we would not need this special case in the description.

Reply:
Ah, I see, the transpose codec would work similarly to the endianness codec in the other PR #155 then. I personally like the `chunk_memory_layout` field more, but the codec seems fine as well.